luskan/src/main/java/org/opencadc/luskan/NumCopiesNavigator.java

```diff
@@ -0,0 +1,144 @@
+ *  License, or (at your option)         licence, soit (à votre gré)
+ *  any later version.                   toute version ultérieure.
+ *
+ *  OpenCADC is distributed in the       OpenCADC est distribué
+ *  hope that it will be useful,         dans l’espoir qu’il vous
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ *  : 5 $
+ *
+ ************************************************************************
+ */
+
+package org.opencadc.luskan;
+
+import ca.nrc.cadc.tap.parser.ParserUtil;
+import ca.nrc.cadc.tap.parser.navigator.FromItemNavigator;
+import ca.nrc.cadc.tap.parser.navigator.ReferenceNavigator;
+import ca.nrc.cadc.tap.parser.navigator.SelectNavigator;
+import java.util.ArrayList;
+import java.util.List;
+import net.sf.jsqlparser.expression.Expression;
+import net.sf.jsqlparser.schema.Table;
+import net.sf.jsqlparser.statement.select.PlainSelect;
+import net.sf.jsqlparser.statement.select.SelectItem;
+import org.apache.log4j.Logger;
+
+/**
+ * Checks there is a single Artifact table in the FROM clause,
+ * then applies the InventoryFunctionConverter to the SELECT, WHERE, and GROUP BY clauses.
+ *
+ */
+public class NumCopiesNavigator extends SelectNavigator {
+    private static final Logger log = Logger.getLogger(NumCopiesNavigator.class);
+
+    InventoryFunctionConverter inventoryFunctionConverter;
+
+    public NumCopiesNavigator(InventoryFunctionConverter en, ReferenceNavigator rn, FromItemNavigator fn) {
+        super(en, rn, fn);
+        this.inventoryFunctionConverter = en;
+    }
+
+    @Override
+    public void visit(PlainSelect ps) {
+        log.debug("visit(PlainSelect) " + ps);
+
+        List<Table> tables = new ArrayList<>();
+        List<Table> fromTableList = ParserUtil.getFromTableList(ps);
+        for (Table fromTable : fromTableList) {
+            if (fromTable.getName().equalsIgnoreCase("Artifact")) {
+                tables.add(fromTable);
+                log.debug("found fromTable: " + fromTable.getWholeTableName());
+            }
+        }
+
+        // if no artifact tables found, return.
```

Comment:
Updated, and renamed, the navigator to simply pass the list of tables in the from to the converter, then apply the converter to the SELECT, WHERE, and GROUP BY. The logic is in the converter, to check if the function can be applied to the query, and throw an exception if it can not.